mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+
+  auto threadViewVal =
+      createThreadView(adaptor.getIn(), blockSize, gridSize, loc, rewriter);
+  ArrayAttr sourceTransformsFromOp;
+  Value source;
+  std::tie(source, sourceTransformsFromOp) =
+      untransform(rewriter, threadViewVal);
+  auto loadOobDims = computeOobFromTransforms(rewriter, sourceTransformsFromOp);
+
+  auto threadViewShape = threadViewVal.getType().cast<ShapedType>().getShape();
+  SmallVector<int64_t> bounds(threadViewShape.size(), 1LL);
+  bounds[threadViewShape.size() - 1] =
+      threadViewShape[threadViewShape.size() - 1];
+  SmallVector<int64_t> strides(threadViewShape.size(), 1LL);
+
+  // Get current workgroup ID.
+  auto bid = rewriter.create<WorkgroupIdOp>(loc, rewriter.getIndexType());
+  // Get current workitem ID.
+  auto tid = rewriter.create<WorkitemIdOp>(loc, rewriter.getIndexType());
+  auto zeroConstantOp = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  SmallVector<Value, 3> loadStartCoords = {bid, tid, zeroConstantOp};
+
+  SmallVector<Value> zeroes(threadViewShape.size(), zeroConstantOp);
+
+  auto outLoop = rewriter.create<TransformingForOp>(
+      loc, ArrayRef<ValueRange>{loadStartCoords},
+      ArrayRef<Attribute>{sourceTransformsFromOp}, ArrayRef<int64_t>(bounds),
+      ArrayRef<int64_t>(strides),
+      /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(outLoop.getBody());
+    auto loadCoords = outLoop.getLowerCoords(/*domain=*/0);
+    auto loadVal = rewriter.create<GlobalLoadOp>(
+        loc, adaptor.getIn().getType().cast<ShapedType>().getElementType(),
+        adaptor.getIn(), std::get<0>(loadOobDims), std::get<1>(loadOobDims),
+        loadCoords);
+    SmallVector<Value, 4> storeCoords;
+    for (const auto &idxAndVal : llvm::enumerate(loadCoords)) {
+      auto idx = idxAndVal.index();
+      auto val = idxAndVal.value();
+      if (idx == redAxis) {
+        storeCoords.push_back(zeroConstantOp);
+      } else {
+        storeCoords.push_back(val);
+      }
+    }
+    StoreMethod stMethod;
+    ReduceMethod rMethod = adaptor.getReduceMethod();
+    if (getStoreMethod(rMethod, stMethod).failed()) {
+      return op.emitError()
+             << "The Reduce Method" << getNameForReduceMethod(rMethod)
+             << " is not supported.!";
+    }
+    rewriter.create<BufferStoreOp>(
+        loc, loadVal, adaptor.getOut(), std::get<0>(loadOobDims),
+        std::get<1>(loadOobDims), storeCoords,
+        StoreMethodAttr::get(rewriter.getContext(), stMethod),
+        /*offset=*/IntegerAttr());
+  }
+  auto region = op->getParentRegion();
+  rewriter.eraseOp(op);
+  (void)runRegionDCE(rewriter, {*region});
+  return success();
+}
+
+void RockLowerReducePass::runOnOperation() {
+  MLIRContext *ctx = &getContext();
+  ConversionTarget target(*ctx);
+
+  target.addIllegalOp<rock::ReduceOp>();
+  target.addLegalOp<rock::TransformOp, rock::TransformingForOp,
```

Comment:
Just `addLegalDialect rock::RockDialect`. Much less painful and won't really cause bugs.

Reply:
done